### ISSUE TITLE
Add support for custom user table

### ISSUE THREAD

**Issue by xperseguers** [codeFareith/cf_google_authenticator#12]
In my project I am centralizing the management of users into a custom "members" table and I hook into the authentication process (both Backend and Frontend) to check credentials against my custom members table and create/update TYPO3 users dynamically, a bit like when dealing with LDAP authentication.

I would like to add support for 2FA.

**Solution**

- Extend my own `ext_tables.sql` to include your 2 database fields
- Override my members TCA with something like:

```
if (ExtensionManagementUtility::isLoaded('cf_google_authenticator')) {
    \call_user_func(
        function () {
            ExtensionManagementUtility::addTCAcolumns(
                'tx_myext_domain_model_member',
                [
                    'tx_cfgoogleauthenticator_enabled' => [
                        'exclude' => true,
                        'label' => PathUtility::makeLocalLangLinkPath(
                            'be_users.tx_cfgoogleauthenticator_enabled',
                            'locallang_db.xlf'
                        ),
                        'config' => [
                            'type' => 'check'
                        ]
                    ],
                    'tx_cfgoogleauthenticator_secret' => [
                        'exclude' => true,
                        'label' => PathUtility::makeLocalLangLinkPath(
                            'be_users.tx_cfgoogleauthenticator_secret',
                            'locallang_db.xlf'
                        ),
                        'config' => [
                            'type' => 'user',
                            'userFunc' => UserSettings::class . '->createSecretField'
                        ]
                    ]
                ]
            );

            ExtensionManagementUtility::addToAllTCAtypes(
                'tx_myext_domain_model_member',
                'tx_cfgoogleauthenticator_enabled,tx_cfgoogleauthenticator_secret',
                '',
                'after:password' // Add the 2FA after our custom field "password"
            );
        }
    );
}
```

This effectively shows the 2FA fields. Now, in order to work a bit further and prevent anyone from disabling 2FA for some arbitrary user w/o providing a proper code, we need to extend your method `\CodeFareith\CfGoogleAuthenticator\Handler\GoogleAuthenticatorSetupHandler::isUsersTable()` so that the custom members table is considered a "users" table as well.

This is something that can easily be done with a hook there.

I already have a working solution so that the "TCA" part of this feature request is ready. However I know that I will need to somehow invoke your authentication code in my own authentication service and thus I suggest that this ticket is really about implementing support from A to Z and I will possibly suggest some (hopefully) minor additional changes to your extension to support this use case.

**Comment by codeFareith** [comment]
Thank you very much for your input.
I'll let you know when I'm done with implementation :)

**Comment by xperseguers** [comment]
Quite a lot has changed internally and my hook does not work anymore, with all the changes, the implementation of this feature is now not as simple anymore :/

**Comment by codeFareith** [comment]
Hey @xperseguers,

please excuse my delayed response. I just came back from London yesterday ^^"

Just to be sure:
what you're trying to achieve - or what you need - is the possibility to tell the cf_google_authenticator extension to not use "fe_users" and "be_users", but to rely on your custom "users" - or "members" - tables instead.

Not only should the extension be able to recognize your tables as valid user tables, but also use them when reading and persisting a specific user's authenticator settings and during the OTP verification process.

In other words:
The entire business logic of the extension should be decoupled (or at least decouplable) from the TYPO3 core user management.

Did I get that right so far?

**Comment by xperseguers** [comment]
Hello,

Yes that's it.

To make things clear again, when I opened this ticket and created a suggested PR, things were actually quite easy:

* I have an extension which provides a domain model for my members `tx_lionsbase_domain_model_member`
* I have an extension which takes care of authenticating (BE and FE) my members

* In my authentication extension, I override the `tx_lionsbase_domain_model_member` by adding 2FA columns if `EXT:cf_google_authenticator` is present:

```
if (ExtensionManagementUtility::isLoaded('cf_google_authenticator')) {
    \call_user_func(
        function () {
            ExtensionManagementUtility::addTCAcolumns(
                'tx_lionsbase_domain_model_member',
                [
                    'tx_cfgoogleauthenticator_enabled' => [
                        'exclude' => false,
                        'label' => PathUtility::makeLocalLangLinkPath(
                            'be_users.tx_cfgoogleauthenticator_enabled',
                            'locallang_db.xlf'
                        ),
                        'config' => [
                            'type' => 'check'
                        ]
                    ],
                    'tx_cfgoogleauthenticator_secret' => [
                        'exclude' => false,
                        'label' => PathUtility::makeLocalLangLinkPath(
                            'be_users.tx_cfgoogleauthenticator_secret',
                            'locallang_db.xlf'
                        ),
                        'config' => [
                            'type' => 'user',
                            'userFunc' => UserSettings::class . '->createSecretField'
                        ]
                    ]
                ]
            );

            ExtensionManagementUtility::addToAllTCAtypes(
                'tx_lionsbase_domain_model_member',
                'tx_cfgoogleauthenticator_enabled,tx_cfgoogleauthenticator_secret',
                '',
                'after:password' // Add the 2FA after our custom field "password"
            );
        }
    );
}
```

This is the first part that needed your extension to be extended because your hook doing business logic on those fields is checking whether the hook is called for `fe_users` or `be_users` and if not quit right away. This is why I suggested a hook (but a signal is fine as well) so that my extension could tell you "yes, `tx_lionsbase_domain_model_member` is a "user" table as well, please do your business check as well.

For the second part, when authenticating with my own authentication business logic, of course, I need to check 2FA. And this was easy to do as well:

1. Do my own logic to check username + password and create/update the TYPO3 user if needed
2. Check for 2FA enabled (in case of BE login for me at the moment) and if so call your check method:

```
                if (TYPO3_MODE === 'BE' && ExtensionManagementUtility::isLoaded('cf_google_authenticator')) {
                    $userRow = GeneralUtility::makeInstance(ConnectionPool::class)
                        ->getConnectionForTable('tx_lionsbase_domain_model_member')
                        ->select(
                            ['tx_cfgoogleauthenticator_enabled', 'tx_cfgoogleauthenticator_secret'],
                            'tx_lionsbase_domain_model_member',
                            [
                                'uid' => $member->getUid(),
                            ]
                        )
                        ->fetch();

                    if ((bool)$userRow['tx_cfgoogleauthenticator_enabled'] === true) {
                        $otp = GeneralUtility::_GP('google-authenticator-otp');
                        $secret = $userRow['tx_cfgoogleauthenticator_secret'];

                        if (GoogleAuthenticatorUtility::verifyOneTimePassword($secret, $otp) === false) {
                            // Invalid OTP!
                            return false;
                        }
                    }
                }
```

That's it! So really nothing complex.

Unfortunately in the mean time you refactored your extension really a lot and when I tried to update my PR to be in line with your changes, it turned out this got much much much more complex!

**Comment by codeFareith** [comment]
Hey @xperseguers,

just made a new release 1.2.0. Now you should be able to use "collectAllowedTables"-Signal, to add any table you'd like to the verification.

The "collectAllowedTables"-Signal gives you the following array:
```
'tables' => [
    'be_users',
    'fe_users'
],
 'caller' => $this,
```

So just add your tables to 'table' and return it via the Slot-Dispatcher back to the Signal.

Please verify, thanks!

**Comment by xperseguers** [comment]
Hey @codeFareith,

* Custom Backend authentication check still woks fine with version 1.2.0
* Integration with custom record: seems not 100% working, I'm analyzing.

Thanks so far for the follow-up.

**Comment by xperseguers** [comment]
found the bug (there is actually one) and a nice-to-have

**Comment by xperseguers** [comment]
I found another bug in the signal, fixing it. Just as a sidenote, it works but I don't find it really cool to extend. Here's how one is supposed to use your signal:

```
    /**
     * @param array $tables
     * @param GoogleAuthenticatorSetupHandler $pObj
     * @return array
     * @see \CodeFareith\CfGoogleAuthenticator\Handler\GoogleAuthenticatorSetupHandler::isUsersTable()
     */
    public function collectAllowedTables(array $tables, GoogleAuthenticatorSetupHandler $pObj): array
    {
        $tables[] = 'tx_myext_domain_model_member';

        return [
            $tables,
            $pObj,
        ];
    }
```

It works! But as you see, due to the way you invoke and use the signal, you need to return an array with the same input parameters. I'm more used to sending a signal in my code and not use the return at all, but instead pass what I expect to be possibly changed as reference, so table in that case.

Rationale? If we have multiple slots for your signal and I'm the first one to implement it, I could possibly return `null` or another class as second parameter and this would possibly crash the next slot with an unexpected class mismatch.

But just leave it like this for the time being, it's not important.

**Comment by xperseguers** [comment]
Hey @codeFareith please check my PR for fixing a bug in your implementation, even when not extending for custom tables and the other one for "polishing" of the custom table extension.

**Comment by codeFareith** [comment]
Thanks for your effort!
I'm going to close this issue; any other bugs should be discussed in separate tickets.
